build-system/tasks/helpers.js

```diff
@@ -453,42 +476,43 @@ async function compileUnminifiedJs(srcDir, srcFilename, destDir, options) {
     };
   }
 
-  /**
-   * Bundles an entry point with all its imports and applies babel transforms.
-   * @param {boolean} continueOnError
-   */
-  async function performBundle(continueOnError) {
-    const {banner, footer} = splitWrapper();
-    await esbuild
-      .build({
-        entryPoints: [entryPoint],
-        bundle: true,
-        sourcemap: true,
-        define: experimentDefines,
-        outfile: destFile,
-        plugins: [esbuildBabelPlugin],
-        banner,
-        footer,
-      })
-      .catch((err) => handleBundleError(err, continueOnError, destFilename));
-    finishBundle(srcFilename, destDir, destFilename, options);
-    let name = destFilename;
-    if (options.latestName) {
-      const latestMaxName = options.latestName.split('.js')[0] + '.max.js';
-      name = `${name} → ${latestMaxName}`;
-    }
-    endBuildStep('Compiled', name, startTime);
-    const target = path.basename(destFilename, path.extname(destFilename));
-    if (UNMINIFIED_TARGETS.includes(target)) {
-      await applyAmpConfig(
-        destFile,
-        /* localDev */ true,
-        /* fortesting */ options.fortesting
-      );
-    }
-  }
+  const {banner, footer} = splitWrapper();
+  const buildResult = await esbuild
+    .build({
+      entryPoints: [entryPoint],
+      bundle: true,
+      sourcemap: true,
+      define: experimentDefines,
+      outfile: destFile,
+      plugins: [esbuildBabelPlugin],
+      banner,
+      footer,
+      incremental: !!options.watch,
+      logLevel: 'silent',
+    })
+    .then((result) => {
+      finishBundle(srcFilename, destDir, destFilename, options, startTime);
+      return result;
+    })
+    .catch((err) => handleBundleError(err, !!options.watch, destFilename));
 
-  await performBundle(options.continueOnError);
+  if (options.watch) {
+    watchedTargets.set(entryPoint, {
+      rebuild: async () => {
+        const time = Date.now();
+        const buildPromise = buildResult
+          .rebuild()
+          .then(() =>
+            finishBundle(srcFilename, destDir, destFilename, options, time)
+          )
+          .catch((err) =>
+            handleBundleError(err, /* continueOnError */ true, destFilename)
+          );
+        options?.onWatchBuild(buildPromise);
+        await buildPromise;
+      },
+    });
+  }
```

Comment:
I'm wonder if this block should be above the call to `esbuild.build`. Here's why:

The idea behind the previous implementation was to `await` the current build _after_ instantly returning the pending build to the lazy builder so it knows ASAP that there is a build going on.

By making this block live below it appears that you're signaling back to the lazy builder after waiting for a full build + write config + copy to `-latest`, by which time the lazy build might have already triggered a duplicate build.

If I'm right about this, then I think the way to fix is to just create a promise on line 480, then call this block, and only then `await` the promise created on line 480.

Reply:
It is confusing, but and I'd love to untangle this further.

- All async functions return their promises instantly, and it is up to the caller to determine whether or not to block.
- The old code did not call `onWatchBuild` on initial calls, only on rebuilds. I retained that behavior here.
- I technically could register the `rebuild()` function before completing the initial `esbuild.build()`, but since the caller of this func always calls await on the result (and must), there is no functional difference.

Reply:
I think this sentence addresses the hypothetical concern I had:

> All async functions return their promises instantly, and it is up to the caller to determine whether or not to block.

Carry on!